core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactoryImpl.java

```diff
@@ -347,7 +350,7 @@ public RelDataType createTypeWithNullability(
    * @throws NullPointerException if type is null
    */
   protected RelDataType canonize(final RelDataType type) {
-    return CACHE.getUnchecked(type);
```

Comment:
The Oracle doc says that an soft reference object will be garbage collected when it is "softly reachable":
> Suppose that the garbage collector determines at a certain point in time that an object is softly reachable.

So do you mean the guava CACHE with same soft reference k-v objects is not softly reachable ? Sorry i'm not that familiar with the Java garbage collection mechanism.

[1] https://docs.oracle.com/javase/8/docs/api/java/lang/ref/SoftReference.html
[2] https://docs.oracle.com/javase/8/docs/api/java/lang/ref/package-summary.html#reachability

Reply:
Even though CACHE's value is soft referenced, key is strong referenced, key and value are referencing the same object. So it will never be garbage collected.

Reply:
Can we give a test case to validate that the cache of the key is not strong reference ?

Reply:
We can, but I don't think it is worth the effort.

Reply:
I also that there is another issue https://github.com/apache/calcite/commit/de9a7164b2d3e5a1637254fd0d54c7a4c886166e which is also fixed by you, so the the CACHE pattern seems very much prone to misused, can we have some protection that on other misuse code happens again ?

Reply:
It is indeed error prone. I checked all the other usages, all seem good to me. But it is almost impossible to prevent people from writing that code in any place, unless code reviewers are aware of the risk.